extensions/binding/org.eclipse.smarthome.binding.coapconnectedlighting/src/main/java/org/eclipse/smarthome/binding/coapconnectedlighting/handler/CoAPConnectedLightingHandler.java

```diff
@@ -0,0 +1,361 @@
+
+    private ChannelUID stringChannelTemperatureUID;
+    private ChannelUID stringChannelHumidityUID;
+    private ChannelUID stringChannelPressureUID;
+
+    ScheduledFuture<?> refreshTemperatureJob;
+    ScheduledFuture<?> refreshHumidityJob;
+    ScheduledFuture<?> refreshPressureJob;
+
+    private BigDecimal refresh;
+
+    CoapClient coapClient = new CoapClient();
+
+    String COAP_SERVER_IP_ADDRESS = "0.0.0.0";
+    int COAP_SERVER_PORT = 5683;
+
+    /*
+     * String COAP_SERVER_IP_ADDRESS_LIGHT = "192.168.1.103";
+     * String COAP_SERVER_PORT_LIGHT = "5683";
+     */
+    public CoAPConnectedLightingHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+
+        // Device Configuration Check
+        String deviceIPAddress = (String) this.getConfig().get("ipAddress");
+        int devicePortNumber = -1;
+
+        if (((BigDecimal) this.getConfig().get("portNumber")) != null) {
+            devicePortNumber = ((BigDecimal) this.getConfig().get("portNumber")).intValue();
+        }
+
+        if ((deviceIPAddress != null) && (devicePortNumber != -1)) {
+            if ((!deviceIPAddress.isEmpty()) && (devicePortNumber != -1)) {
+                COAP_SERVER_IP_ADDRESS = deviceIPAddress;
+                COAP_SERVER_PORT = devicePortNumber;
+            } else {
+                logger.error(this.getThing().getUID() + "-Invalid IP Address or Port Number");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Can not access device as the ip address or port number is invalid");
+                updateStatus(ThingStatus.OFFLINE);
+                return;
+            }
+        } else {
+            logger.error(this.getThing().getUID() + "-NULL Configuration - IP Address or Port number is missing");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Can not access device as the ip address or port number is not set (NULL)");
+            updateStatus(ThingStatus.OFFLINE);
+            return;
+        }
+
+        stringChannelTemperatureUID = new ChannelUID(getThing().getUID(), CHANNEL_Temperature);
+        stringChannelHumidityUID = new ChannelUID(getThing().getUID(), CHANNEL_Humidity);
+        stringChannelPressureUID = new ChannelUID(getThing().getUID(), CHANNEL_Pressure);
+
+        refresh = new BigDecimal(60);
+
+        // Get the device properties
+        Map<String, String> deviceProperties = editProperties();
+
+        // Add the IP Address property
+        deviceProperties.put("IP Address", COAP_SERVER_IP_ADDRESS);
+
+        // Get the device functionality
+        coapClient.setURI("coap://" + COAP_SERVER_IP_ADDRESS + ":" + COAP_SERVER_PORT);
+        Set<WebLink> weblink = coapClient.discover();
+
+        if (weblink != null) {
+            Iterator<WebLink> links = weblink.iterator();
+
+            StringBuilder weblinkString = new StringBuilder();
```

Comment:
It seems the StringBuilder is only there to check if any of the WebLinks provided by the client contain "actuators" or "sensors". You can achieve the same with streams and shorten the code dramatically:

```
// map every WebLink to its URI and return true if any of them URIs contain the given String:
boolean isActuator = weblink.stream().map(WebLink::getURI).anyMatch(uri -> uri.contains("actuators"));
boolean isSensor = weblink.stream().map(WebLink::getURI).anyMatch(uri -> uri.contains("sensors"));
```

This way you can condense the `deviceProperties#put` block to

```
deviceProperties.put("Is a Light ?",  isActuator ? "Yes" : "No");
deviceProperties.put("Is a Sensor ?", isSensor ? "Yes": "No");
```

Reply:
@htreu Thanks for this suggestion. It does shrink the code quite a lot !